### ISSUE TITLE
VSCode analyses `import { sqlite3 } from "sqlite3"` differently, and wrongly, from any other import in JS mode.

### ISSUE THREAD

**Issue by Pomax** [microsoft/vscode#128919]
Does this issue occur when all extensions are disabled?: Yes/No

Versions:

![image](https://user-images.githubusercontent.com/177243/126046994-a80f01ec-155b-45ca-8c1f-8c298bce112c.png)

Problem:

![image](https://user-images.githubusercontent.com/177243/126046988-4e9e9b77-bfaf-4601-8f0c-472579e34ae8.png)

the import is very clearly used, yet VSCode refuses to acknowledge it is. No matter how immediately I use it, VSCode keeps complaining:

![image](https://user-images.githubusercontent.com/177243/126047022-48d1f480-53c7-408b-837f-712ec914d9c8.png)

**Comment by Pomax** [comment]
Note: changing this to `import { sqlite3 } from "better-sqlite3"` or literally anything else works, so VSCode somehow applies _different logic_ for when something specifically imports from `"sqlite3"`, and it definitely shouldn't.

**Comment by mjbvz** [comment]
Please share a minimal example project that demonstrates the issue

Also please fill in the issue template /needsMoreInfo

**Comment by Pomax** [comment]
There is literally nothing to example: open VSCode, type `import { sqlite3 } from "sqlite3"; sqlite3.doAThing();` and set the file to javascript. Done: it immediately shows off this problem.

**Comment by thebinarysearchtree** [comment]
This is not specific to sqlite3. There are two types of NodeJS modules - CommonJS and es6. CommonJS are the older type of module and they can only be imported with the default import syntax, which is:

`import sqlite3 from 'sqlite3';`

**Comment by Pomax** [comment]
No, it's not: what you're showing is a full namespace import (e.g. a module's `export default`, or full `export` map), whereas what I'm showing is an import of an explicitly named export. This is normal ES module syntax, and VSCode is doing the wrong thing on it specifically and only when the import is for the string "sqlite3". It doesn't do this for _any_ other import.

**Comment by thebinarysearchtree** [comment]
Yes, I don't know why it is saying that the import isn't used, that is something to do with TypeScript. It might be because VSCode uses sqlite3 itself and it is interacting in a strange way.

What I am saying though is that your import statement isn't correct. NodeJS can sometimes get named imports right, but the only reliable way to import a CommonJS modules is with the syntax I mentioned. Even if you wanted to do a named import, you don't seem to be doing it correctly. I assume you want to use the Database class. In that case, it would be:

`import { Database } from 'sqlite3';`

TypeScript has a bunch of cached types, including sqlite3 types. The export called "sqlite3" is actually an interface, so I don't know how that corresponds to JavaScript.

**Comment by thebinarysearchtree** [comment]
It knows what sqlite3 is because you specifically mentioned it in the import statement. It then looks up its type definitions for that module, which can only refer to the NodeJS module called sqlite3 because there can only be one. NodeJS does not reliably import CommonJS modules with that syntax. Even if you wanted to use a named import, the documentation seems to want you to use either the class called Database, or the function called verbose. That would be:

`import { Database } from 'sqlite3';`

It probably won't work at run-time though.

**Comment by thebinarysearchtree** [comment]
1. VSCode uses TypeScript to analyse JavaScript. 
2. Your import statement references the npm package called sqlite3. 
3. TypeScript uses DefinitelyTyped to find the type definitions of sqlite3. 
4. It doesn't matter what is in your package.json or if it exists or not. 
5. sqlite3's documentation refers to two different imports - Database and verbose. 
6. You are using a named import of sqlite3, which is neither of these. 
7. DefinitelyTyped includes a definition for sqlite3, but it is an interface that is not meant to be used directly. 
8. NodeJS can only reliably import CommonJS modules using the default import syntax.

**Comment by thebinarysearchtree** [comment]
Yup, you are wrong.

`
import * as sqlite from 'sqlite3';
import sqlite3 from 'sqlite3';
import { Database } from 'sqlite3';
`

I don't know if the last one will work or not. That can only be determined at run-time.

If you think you are right, it is very easy to prove - run the program.

**Comment by thebinarysearchtree** [comment]
> No, it's not

Yes, it is, and you can't easily find this out by running your program, if you ever finish writing it. You will need to change it to the default import syntax.

**Comment by wangweixuan** [comment]
You did import it in a wrong way.

According to the [type definition](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4dc799d17c4abd27a1c490d959aaabc9c47869e5/types/sqlite3/index.d.ts#L96), the named export called `splite3` is an *interface*, that means it's not a concrete thing in JavaScript. Interfaces only exist in compile time, and only in TypeScript. However, you are importing it in a JavaScript file, which is meaningless! It's reported as *unused* because you can't use something that doesn't exist!

The correct way is:
```js
import { DataBase } from "sqlite3";
new DataBase(...);
// or
import * as sqlite3 from "sqlite3";
new sqlite3.DataBase(...);
```

I recommend you to turn on `checkJs` in `tsconfig.json`. Then it will report an error as follows and help you find the problem:

> 'sqlite3' only refers to a type, but is being used as a value here. ts(2693)

**Comment by Pomax** [comment]
Again, no: because I'm not using typescript, or node, or anything else: we're talking about pure, spec-compliant Ecmascript 262 v6 module syntax. VSCode makes assumptions about what the "sqlite3" module is when it shouldn't, because in the absence of any information other than "this is JS code", the _syntax shows_ should, for a spec-compliant parser, lead to the conclusion that the import from the module is in scope by the second line, and nothing is "imported but not used". Given the [official Ecmascript 262 v6 specification for the import statement](https://tc39.es/ecma262/#sec-imports), VSCode is doing the wrong thing here.

Do with that as you will, I've given you all the information several times now.
